### PR TITLE
Add Profile template argument to dispatcher.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44033 DeviceGuard dispatch key
* **#44053 Add Profile template argument to dispatcher.**

The profile template argument can be used to prevent generation of
profiling code for a dispatch call, which is a good idea when your
dispatch handler is very small and you shouldn't spend time profiling
it.

While doing this, I also restructured redispatch so it also was
a method on TypedOperatorHandle, and rewrote some call sites to use
the direct method, which shortened some of our generated code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
This is probably not actually a good idea, as it has unexpected interactions with `shouldRecord` call inside the record function call

**Comment by dzhulgakov** [comment]
What were you trying to achieve? In my understanding we should profile any user-initiated dispatch invocation at most once (even if it internally does redispatch). So the fact that we need this trickery probably hints that our profiling code is not in the right place and/or there's no distinction between user-level entrance to dispatcher vs re-dispatching.

**Comment by ezyang** [comment]
> So the fact that we need this trickery probably hints that our profiling code is not in the right place and/or there's no distinction between user-level entrance to dispatcher vs re-dispatching.

@dzhulgakov  IMO, the profiling code isn't in the right place. @bhosmer and I talked about this when profiling integration initially landed; my strong belief is that the profiling code makes the most semantic sense when placed on the kernels themselves, rather than in the dispatcher. Then I can just omit the profiling code when I have cheap wrappers that I don't want to profile. With the current system, I have to make the runtime test in `shouldRecord` more complicated to do this.

I tried to see if I could write a template wrapper to insert profiling in kernels automatically, and I got part of the way there, but there are some barriers:

```
diff --git a/torch/library.h b/torch/library.h
index 5c8936b6b8..bd368bc89b 100644
--- a/torch/library.h
+++ b/torch/library.h
@@ -74,6 +74,41 @@ namespace torch {
 template <class CurClass>
 class class_;
 
+template<class FuncTy, FuncTy* Func>
+struct profile_wrapper_impl final {
+};
+
+template<class Ret, class... Args, Ret (*Func)(Args...)>
+struct profile_wrapper_impl<Ret(Args...), Func> final {
+  static decltype(auto) wrapper(Args... args) {
+    at::RecordFunction guard(at::RecordScope::FUNCTION);
+    if (C10_UNLIKELY(guard.active)) {
+      int64_t seq_num = -1;
+      /*
+      // TODO
+      // Setting sequence number in the Autograd case to associate
+      // the forward range with the coresponding Autograd's node
+      if (isIncludedInAlias(dispatchKey, DispatchKey::Autograd) && at::GradMode::is_enabled()) {
+        seq_num = at::sequence_number::peek();
+      }
+      */
+      if (guard.needs_inputs) {
+        torch::jit::Stack stack = c10::impl::BoxedKernelWrapper<Ret(Args...)>::boxArgs(args...);
+        // TODO: name of the function
+        guard.before("", stack, seq_num);
+      } else {
+        guard.before("", seq_num);
+      }
+    }
+    return (*Func)(std::forward<Args>(args)...);
+  }
+};
+
+template<typename FuncPtr>
+constexpr auto profile_wrapper(FuncPtr kernel_func) {
+  return &profile_wrapper_impl<typename FuncPtr::FuncType, FuncPtr::func_ptr()>::wrapper;
+}
+       
 /// Represents a C++ function that implements an operator.  Most users won't
 /// interact directly with this class, except via error messages: the
 /// constructors this function define the set of permissible "function"-like
```

The big problem is knowing what name of the function is when doing profiling; I can't get this information from the current registration API, and so the only way I could think of that preserved the current registration API was to change the unboxed calling convention so that the operator handle is passed through by default too (not the *worst* idea in the world, it would help us do some other stuff too).

**Comment by dzhulgakov** [comment]
Hm, that wrapper is not ideal either: more wrappers, probably means more code bloat and overhead.

Can there be a separation in Dispatcher between "real entrance" and "redispatch"? Or is it what this PR was trying to do?

**Comment by ezyang** [comment]
> Can there be a separation in Dispatcher between "real entrance" and "redispatch"? Or is it what this PR was trying to do?

Yes, this was this PR was trying to do. But I can't do it as a dispatcher level API, because profiling decisions are made by looking at the dispatch key *we are about to dispatch to* (`shouldRecord`) and then excluding dispatch based on this situation.

We could also do the old design I argued for in https://github.com/pytorch/pytorch/pull/37587#discussion_r444300175

> It seems clear you don't want to profile twice in this case. But there should be a very simple fix for this: make BackendSelect redispatch bypass record function. Then you can avoid this special case entirely.

But when @bhosmer and I discussed it, it seemed clear that most people think about profiling as a caller (not callee affair): you profile a section of code you are about to execute, as opposed to you get profiled by whoever calls you. The wrapper encodes this more clearly.

By the way, if the wrapper is inlined, I think the net code bloat might be a wash. When you inline into kernels, your cost is number of kernels (which is operators x backends); when you put it on the dispatch site, your cost is number of dispatch sites (operators + all other direct calls to dispatcher, which includes most redispatch sites).

**Comment by ezyang** [comment]
cc @ilia-cher

**Comment by ilia-cher** [comment]
> cc @ilia-cher

will comment today

File: aten/src/ATen/core/dispatch/Dispatcher.h [hunk 1]
```diff
@@ -396,25 +407,28 @@ inline Return Dispatcher::redispatch(const TypedOperatorHandle<Return (Args...)>
...
-      if (isIncludedInAlias(dispatchKey, DispatchKey::Autograd) && at::GradMode::is_enabled()) {
-        seq_num = at::sequence_number::peek();
-      }
-      if (guard.needs_inputs) {
-        guard.before(op.schema().name(), *stack, seq_num);
-      } else {
-        guard.before(op.schema().name(), seq_num);
+  if (Profile) {
+    // using already existing stack to record function execution in observers
+    at::RecordFunction guard(at::RecordScope::FUNCTION);
```
Comment (ilia-cher) [inline aten/src/ATen/core/dispatch/Dispatcher.h:420]: i think this just destroys the guard before the kernel is executed, `kernel.callBoxed(op, stack);` then should be in the same scope as the guard

File: aten/src/ATen/core/dispatch/Dispatcher.h [hunk 2]
```diff
@@ -337,53 +346,55 @@ namespace detail {
...
-      } else {
-        guard.before(op.schema().name(), seq_num);
+  if (Profile) {
+    // Check if we need to run callbacks registered with RecordFunction
+    // If true and callbacks need inputs, we box the arguments and pass
+    // them into the callbacks and also into the kernel call
+
+    // Note: for perf reasons we wouldn't want to pass arguments into
+    // the function call or prematurely box them
+    at::RecordFunction guard(at::RecordScope::FUNCTION);
```
Comment (ilia-cher) [inline aten/src/ATen/core/dispatch/Dispatcher.h:364]: same here, kernel call should be in the same scope as the guard

**Review by ilia-cher — CHANGES_REQUESTED** [review]
it seems there's an issue with the guards placement

I remember the concern that we might lose some ops that don't use codegen(?) if we move profiling code into the op wrappers (c++ registered c10 ops? are there other cases?)

btw what are the places where you'd want to explicitly call with Profile set to false? is that to remove `isObserved` checks?

**Comment by ezyang** [comment]
@ilia-cher the intention of this PR is not to merge it / do CR on it; instead, Dmytro and I were discussing some general high level design based on this PR. (Though, I agree there is a problem with this code!) Would you mind checking over the existing thread and weighing in?

**Comment by ilia-cher** [comment]
> @ilia-cher the intention of this PR is not to merge it / do CR on it; instead, Dmytro and I were discussing some general high level design based on this PR. (Though, I agree there is a problem with this code!) Would you mind checking over the existing thread and weighing in?

Sure, generally I'm not against moving RecordFunction; my prev. comment though:

> I remember the concern that we might lose some ops that don't use codegen(?) if we move profiling code into the op wrappers (c++ registered c10 ops? are there other cases?)

I guess, one thing I'd want to understand better is where exactly in the codegen we'd want to put RecordFunction. As I remember from the prev. discussions there were multiple options, and depending on it some ops might not be covered - this might be ok, but I'd want to better understand the options.

**Comment by ezyang** [comment]
> I guess, one thing I'd want to understand better is where exactly in the codegen we'd want to put RecordFunction. As I remember from the prev. discussions there were multiple options, and depending on it some ops might not be covered - this might be ok, but I'd want to better understand the options.

In the patch I posted above, the suggestion was to do it as a template wrapper applied by the op registration API, which means by default you would get RecordFunction on all registered functions (and then we'd have special registration API to bypass this). That would make it easier to make sure most people get this coverage. However, this is difficult to implement as we can't easily recover the operator name via a template wrapper like this without changing the unboxed calling convention.

**Comment by ilia-cher** [comment]
are there ops that don't go through the op registration API?
in any case, we do need name and arguments, the exact place of RecordFunction can be changed as long as we capture the op invocations (once per invocation), and we have op's name and arguments

**Comment by dzhulgakov** [comment]
Have we considered a slight twist on your templated kernel wrappers? I.e. can we keep ProfiledType (still handled by dispatcher in generic fashion) but make it's bit to be determined by hardcoded codepath in dispatcher.

Basically dispatcher code would look something like this:
```
Dispatcher::call() {
  keyset = determine_dispatch_keys(...)
  if (keyset.lowest() == Profiling) {
    bool shouldProfile = toss_the_coin_and_the_logic_of_record_function_to_decide_whether_to_run_or_not();
    if (!shoudProfile) keyset.remove(Profiling);
  }
  // regular dispatch to `keyset.lowest()`
}

// this can be codegened or can be even boxed fallback
ProfilingType::foo() {
  RecordFunction rf(<some context on the tossed coin, can be passed through threadlocal if needed>);
  Dispatcher::redispatch()
}
```

This approach can combine the best parts of the current state of code and your previous state:
- there's no overhead in the no-profiling mode beyond just tossing the coin
- there's even less code bloat because each invocation of Dispatcher will not include the boxing code to call profiler
- profiling code has some overhead, but it's fine
- no need to change op registration api or template wrappers there (as it's more complexity)
- we can even try to have a single profiling boxed fallback and not have ProfilingType

wdyt?

**Comment by ilia-cher** [comment]
as I remember, the issue with ProfiledType dispatch key was that is was set thread locally and then we could not normally use global observers

**Comment by ezyang** [comment]
> are there ops that don't go through the op registration API?

No, everything goes through the op registration API!

> as I remember, the issue with ProfiledType dispatch key was that it was set thread locally and then we could not normally use global observers

This is right. When there are global observers, ProfiledType will typically always be in the TLS.

Still, there may be some merit to this approach. In particular, this is an OK strategy if we want profiling to only happen once per dispatch chain, which is what we used to do. We thought the more detailed profiling was a benefit when reviewing the original RecordFunction patch, but if performance is a problem we could go less detailed.

**Comment by ilia-cher** [comment]
> ProfiledType will typically always be in the TLS.

In practical scenarios unfortunately we noticed that we can't always guarantee that the TLS flag is set or propagated correctly. 
I think we've had examples of this in production when logging or tracing did not work in some of workflows. Even if we e.g. set it in torch module init, it's hard to guarantee that the flag is not lost when model ops/methods start running. That was the main reason to move away from using TLS flag back then. Generally speaking, thread local observers don't work too well with process-level fleet-wide tracing/logging that we use in prod.

**Comment by dzhulgakov** [comment]
I wasn't actually suggesting to propagate TSL profiling flag. What I was suggesting is to still have a bit in the keyset and use it for dispatcher machinery. But this bit would be special - every time when we compute the keyset we'd toss a coin and determine this bit's value. That way we can likely avoid a lot of extra templated code because when the bit is on, it'd be a regular redispatch codepath.

**Comment by ilia-cher** [comment]
> I wasn't actually suggesting to propagate TSL profiling flag. What I was suggesting is to still have a bit in the keyset and use it for dispatcher machinery. But this bit would be special - every time when we compute the keyset we'd toss a coin and determine this bit's value. That way we can likely avoid a lot of extra templated code because when the bit is on, it'd be a regular redispatch codepath.

so `determine_dispatch_keys` in your code snippet always makes sure to set Profiling key?

**Pull request closed without merging**